File: tools/p11rand-seed.c
/*
 * p11rand-seed — seed the Linux kernel RNG from a PKCS#11 module.
 *
 * Opens the module, picks the first usable slot (which is the configured
 * pool), opens a session, and then periodically calls C_GenerateRandom
 * for N bytes and pushes those bytes into /dev/random. By default the
 * insertion uses the RNDADDENTROPY ioctl, which both adds the bytes and
 * credits N*8 bits of entropy to the kernel pool — that requires
 * CAP_SYS_ADMIN (typically root). Pass --no-credit to fall back to a
 * plain write(2), which still mixes the bytes in but credits no
 * entropy; useful for non-root testing or as a hedge.
 *
 * The pool slot must already be present at startup; pair this with
 * p11rand-wait if your sources may not be ready yet (e.g. a systemd
 * unit ordered after card insertion).
 *
 * After each successful RNDADDENTROPY the seeder also issues the
 * RNDRESEEDCRNG ioctl, forcing the kernel's CRNG to mix the newly
 * credited entropy in right away rather than waiting for its next
 * scheduled reseed. Both ioctls share the CAP_SYS_ADMIN requirement;
 * either falls back gracefully if the kernel returns EPERM.
 *
 * Usage:
 *     p11rand-seed [-t SECONDS] [-n BYTES] [--no-credit] [--no-reseed] [-q] <module.so>
 *
 *       -t SECONDS    interval between insertions; fractional OK; default 60
 *       -n BYTES      bytes per insertion; 1..4096; default 64
 *       --no-credit   write(2) without entropy credit (skips RNDADDENTROPY)
 *       --no-reseed   skip the RNDRESEEDCRNG follow-up after each credit
 *       -q            suppress informational stderr output
 *
 * Exit status:
 *     0  clean exit on SIGINT / SIGTERM
 *     1  fatal runtime failure
 *     2  command-line / module-load error
 */

#define _POSIX_C_SOURCE 200809L
#define _DEFAULT_SOURCE   /* for explicit_bzero(3) */

#include "../include/pkcs11.h"

#include <dlfcn.h>
#include <errno.h>
#include <fcntl.h>
#include <getopt.h>
#include <signal.h>
#include <stdarg.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include <unistd.h>

#include <sys/ioctl.h>
#include <linux/random.h>

#define DEFAULT_INTERVAL 60.0
#define DEFAULT_NBYTES   64
#define MAX_NBYTES       4096

static int g_quiet, g_no_credit, g_no_reseed;
static volatile sig_atomic_t g_stop;
static void on_signal(int sig) { (void)sig; g_stop = 1; }

static void info(const char *fmt, ...)
{
    if (g_quiet) return;
    va_list ap;
    va_start(ap, fmt);
    vfprintf(stderr, fmt, ap);
    va_end(ap);
}

static void usage(const char *argv0)
{
    fprintf(stderr,
        "usage: %s [-t SECONDS] [-n BYTES] [--no-credit] [--no-reseed] [-q] <module.so>\n"
        "  -t SECONDS    interval between insertions (fractional OK); default %.0f\n"
        "  -n BYTES      bytes per insertion (1..%d); default %d\n"
        "  --no-credit   plain write(2) instead of RNDADDENTROPY (no root needed)\n"
        "  --no-reseed   skip RNDRESEEDCRNG after each successful credit\n"
        "  -q            quiet; no informational stderr output\n",
        argv0, DEFAULT_INTERVAL, MAX_NBYTES, DEFAULT_NBYTES);
}

/* nanosleep that wakes early on g_stop. */
static void interruptible_sleep(double seconds)
{
    if (seconds <= 0) return;
    struct timespec rem, ts = {
        .tv_sec  = (time_t)seconds,
        .tv_nsec = (long)((seconds - (double)(time_t)seconds) * 1e9),
    };
    while (!g_stop && nanosleep(&ts, &rem) < 0 && errno == EINTR)
        ts = rem;
}

int main(int argc, char **argv)
{
    double interval = DEFAULT_INTERVAL;
    long   nbytes   = DEFAULT_NBYTES;

    static struct option longopts[] = {
        { "no-credit", no_argument, NULL, 1 },
        { "no-reseed", no_argument, NULL, 2 },
        { "help",      no_argument, NULL, 'h' },
        { 0, 0, 0, 0 },
    };
    int opt;
    while ((opt = getopt_long(argc, argv, "t:n:qh", longopts, NULL)) != -1) {
        switch (opt) {
        case 't': {
            char *end = NULL;
            interval = strtod(optarg, &end);
            if (end == optarg || interval <= 0) {
                fprintf(stderr, "invalid -t value: %s\n", optarg);
                return 2;
            }
            break;
        }
        case 'n': {
            char *end = NULL;
            nbytes = strtol(optarg, &end, 10);
            if (end == optarg || nbytes <= 0 || nbytes > MAX_NBYTES) {
                fprintf(stderr, "invalid -n value: %s (1..%d)\n",
                        optarg, MAX_NBYTES);
                return 2;
            }
            break;
        }
        case 'q': g_quiet = 1; break;
        case 1:   g_no_credit = 1; break;
        case 2:   g_no_reseed = 1; break;
        case 'h': usage(argv[0]); return 0;
        default:  usage(argv[0]); return 2;
        }
    }
    if (optind >= argc) { usage(argv[0]); return 2; }
    const char *module = argv[optind];

    void *h = dlopen(module, RTLD_NOW);
    if (!h) {
        fprintf(stderr, "dlopen %s: %s\n", module, dlerror());
        return 2;
    }
    CK_RV (*GetFunctionList)(CK_FUNCTION_LIST_PTR_PTR);
    *(void **)&GetFunctionList = dlsym(h, "C_GetFunctionList");
    if (!GetFunctionList) {
        fprintf(stderr, "dlsym C_GetFunctionList: %s\n", dlerror());
        return 2;
    }
    CK_FUNCTION_LIST_PTR f = NULL;
    if (GetFunctionList(&f) != CKR_OK || !f) {
        fprintf(stderr, "C_GetFunctionList failed\n");
        return 2;
    }
    if (f->C_Initialize(NULL) != CKR_OK) {
        fprintf(stderr, "C_Initialize failed\n");
        return 2;
    }

    CK_ULONG nslots = 0;
    if (f->C_GetSlotList(CK_TRUE, NULL, &nslots) != CKR_OK || nslots == 0) {
        fprintf(stderr, "no usable PKCS#11 slot — pool not fillable yet?\n");
        f->C_Finalize(NULL);
        dlclose(h);
        return 1;
    }
    CK_SLOT_ID *slots = calloc(nslots, sizeof(*slots));
    if (!slots) { f->C_Finalize(NULL); dlclose(h); return 1; }
    if (f->C_GetSlotList(CK_TRUE, slots, &nslots) != CKR_OK) {
        fprintf(stderr, "C_GetSlotList failed\n");
        free(slots); f->C_Finalize(NULL); dlclose(h); return 1;
    }

    CK_SESSION_HANDLE sess = 0;
    if (f->C_OpenSession(slots[0], CKF_SERIAL_SESSION,
                         NULL, NULL, &sess) != CKR_OK) {
        fprintf(stderr, "C_OpenSession failed\n");
        free(slots); f->C_Finalize(NULL); dlclose(h); return 1;
    }
    free(slots);

    int rfd = open("/dev/random", O_WRONLY);
    if (rfd < 0) {
        fprintf(stderr, "open /dev/random: %s\n", strerror(errno));
        f->C_CloseSession(sess); f->C_Finalize(NULL); dlclose(h); return 1;
    }

    /* One contiguous allocation: rand_pool_info header followed by the
     * byte buffer (the trailing flex array). C_GenerateRandom writes
     * straight into rpi->buf, RNDADDENTROPY consumes the same memory. */
    size_t rpi_size = sizeof(struct rand_pool_info) + (size_t)nbytes;
    struct rand_pool_info *rpi = calloc(1, rpi_size);
    if (!rpi) {
        close(rfd); f->C_CloseSession(sess); f->C_Finalize(NULL); dlclose(h);
        return 1;
    }
    rpi->entropy_count = (int)(nbytes * 8);
    rpi->buf_size      = (int)nbytes;

    struct sigaction sa = { .sa_handler = on_signal };
    sigemptyset(&sa.sa_mask);
    sigaction(SIGINT,  &sa, NULL);
    sigaction(SIGTERM, &sa, NULL);

    info("seeding /dev/random: %ld B every %.2f s%s\n",
         nbytes, interval,
         g_no_credit ? " (no entropy credit)" : "");

    int rv = 0;
    while (!g_stop) {
        if (f->C_GenerateRandom(sess,
                                (CK_BYTE_PTR)rpi->buf,
                                (CK_ULONG)nbytes) != CKR_OK) {
            if (g_stop) break;
            fprintf(stderr, "C_GenerateRandom failed; retrying after interval\n");
            interruptible_sleep(interval);
            continue;
        }

        if (g_no_credit) {
            ssize_t w = write(rfd, rpi->buf, (size_t)nbytes);
            if (w < 0) {
                fprintf(stderr, "write /dev/random: %s\n", strerror(errno));
            } else {
                info("wrote %zd B (uncredited)\n", w);
            }
        } else if (ioctl(rfd, RNDADDENTROPY, rpi) < 0) {
            if (errno == EPERM) {
                fprintf(stderr,
                    "RNDADDENTROPY: EPERM — falling back to uncredited write "
                    "(run as root or with CAP_SYS_ADMIN to credit entropy)\n");
                g_no_credit = 1;
                ssize_t w = write(rfd, rpi->buf, (size_t)nbytes);
                if (w < 0)
                    fprintf(stderr, "write /dev/random: %s\n", strerror(errno));
            } else {
                fprintf(stderr, "RNDADDENTROPY: %s\n", strerror(errno));
            }
        } else {
            info("credited %ld B (%d bits) to kernel pool\n",
                 nbytes, rpi->entropy_count);
#ifdef RNDRESEEDCRNG
            if (!g_no_reseed) {
                if (ioctl(rfd, RNDRESEEDCRNG) < 0) {
                    if (errno == EPERM) {
                        fprintf(stderr,
                            "RNDRESEEDCRNG: EPERM — disabling for the run\n");
                        g_no_reseed = 1;
                    } else {
                        fprintf(stderr, "RNDRESEEDCRNG: %s\n", strerror(errno));
                    }
                } else {
                    info("forced CRNG reseed\n");
                }
            }
#else
            (void)g_no_reseed;
#endif
        }

        interruptible_sleep(interval);
    }

    info("exiting\n");

    /* Don't leave the byte buffer in memory after exit. */
    explicit_bzero(rpi, rpi_size);
    free(rpi);
    close(rfd);
    f->C_CloseSession(sess);
    f->C_Finalize(NULL);
    dlclose(h);
    return rv;
}

File: crates/p11rand-tools/src/bin/p11rand-seed.rs
//! `p11rand-seed` — seed the Linux kernel RNG from a PKCS#11 module.
//!
//! Periodically calls `C_GenerateRandom` for N bytes and pushes them into
//! `/dev/random`. By default the insertion uses the `RNDADDENTROPY` ioctl,
//! which both mixes the bytes in *and* credits N*8 bits of entropy to the
//! kernel pool — that requires `CAP_SYS_ADMIN` (typically root). Pass
//! `--no-credit` to fall back to a plain write(2) (no entropy credit).
//!
//! After each successful credit the seeder also issues `RNDRESEEDCRNG`,
//! forcing the kernel CRNG to mix the new entropy in immediately. Both
//! ioctls share the `CAP_SYS_ADMIN` requirement; either falls back
//! gracefully if the kernel returns `EPERM`.

#![allow(non_snake_case)]

use std::ffi::c_int;
use std::fs::OpenOptions;
use std::io::Write;
use std::mem::size_of;
use std::os::unix::fs::OpenOptionsExt;
use std::os::unix::io::AsRawFd;
use std::process::ExitCode;
use std::time::{Duration, Instant};

use p11rand_tools::cryptoki_sys::{
    CKF_SERIAL_SESSION, CKR_OK, CK_SESSION_HANDLE, CK_SLOT_ID, CK_TRUE, CK_ULONG,
};
use p11rand_tools::{install_signals, load_module, signal_count, signal_reset};

const DEFAULT_INTERVAL: f64 = 60.0;
const DEFAULT_NBYTES: usize = 64;
const MAX_NBYTES: usize = 4096;

// `/usr/include/linux/random.h`:
//   #define RNDADDENTROPY  _IOW('R', 0x03, int[2])
//   #define RNDRESEEDCRNG  _IO ('R', 0x07)
// Both ioctl numbers are stable across the kernels we care about.
const RNDADDENTROPY: libc::c_ulong = 0x4008_5203;
const RNDRESEEDCRNG: libc::c_ulong = 0x0000_5207;

#[repr(C)]
struct RandPoolInfoHeader {
    entropy_count: c_int,
    buf_size: c_int,
    // `__u32 buf[0]` flex array follows in memory.
}

struct Args {
    interval: f64,
    nbytes: usize,
    no_credit: bool,
    no_reseed: bool,
    quiet: bool,
    module: String,
}

fn usage(argv0: &str) {
    eprintln!(
        "usage: {argv0} [-t SECONDS] [-n BYTES] [--no-credit] [--no-reseed] [-q] <module.so>"
    );
    eprintln!("  -t SECONDS    interval between insertions (fractional OK); default 60");
    eprintln!("  -n BYTES      bytes per insertion (1..{MAX_NBYTES}); default {DEFAULT_NBYTES}");
    eprintln!("  --no-credit   plain write(2) instead of RNDADDENTROPY (no root needed)");
    eprintln!("  --no-reseed   skip RNDRESEEDCRNG after each successful credit");
    eprintln!("  -q            quiet; no informational stderr output");
}

fn parse() -> Result<Args, ExitCode> {
    let argv: Vec<String> = std::env::args().collect();
    let mut a = Args {
        interval: DEFAULT_INTERVAL,
        nbytes: DEFAULT_NBYTES,
        no_credit: false,
        no_reseed: false,
        quiet: false,
        module: String::new(),
    };
    let mut i = 1;
    while i < argv.len() {
        match argv[i].as_str() {
            "-t" => {
                i += 1;
                let v: f64 = argv
                    .get(i)
                    .and_then(|s| s.parse().ok())
                    .ok_or(ExitCode::from(2))?;
                if v <= 0.0 {
                    return Err(ExitCode::from(2));
                }
                a.interval = v;
            }
            "-n" => {
                i += 1;
                let v: usize = argv
                    .get(i)
                    .and_then(|s| s.parse().ok())
                    .ok_or(ExitCode::from(2))?;
                if v == 0 || v > MAX_NBYTES {
                    return Err(ExitCode::from(2));
                }
                a.nbytes = v;
            }
            "--no-credit" => a.no_credit = true,
            "--no-reseed" => a.no_reseed = true,
            "-q" => a.quiet = true,
            "-h" | "--help" => {
                usage(&argv[0]);
                return Err(ExitCode::from(0));
            }
            s if !s.starts_with('-') => {
                a.module = s.to_string();
                if i + 1 < argv.len() {
                    return Err(ExitCode::from(2));
                }
            }
            _ => {
                usage(&argv[0]);
                return Err(ExitCode::from(2));
            }
        }
        i += 1;
    }
    if a.module.is_empty() {
        usage(&argv[0]);
        return Err(ExitCode::from(2));
    }
    Ok(a)
}

macro_rules! info {
    ($q:expr, $($t:tt)*) => {
        if !$q { eprintln!($($t)*); }
    };
}

fn interruptible_sleep(seconds: f64) {
    if seconds <= 0.0 {
        return;
    }
    // Wake periodically (50ms) so we notice signal_count() going up promptly.
    let total = Duration::from_secs_f64(seconds);
    let start = Instant::now();
    let tick = Duration::from_millis(50);
    while start.elapsed() < total && signal_count() == 0 {
        let remaining = total.saturating_sub(start.elapsed());
        std::thread::sleep(remaining.min(tick));
    }
}

fn main() -> ExitCode {
    let mut a = match parse() {
        Ok(a) => a,
        Err(c) => return c,
    };

    let m = match load_module(&a.module) {
        Ok(m) => m,
        Err(e) => {
            eprintln!("{e}");
            return ExitCode::from(2);
        }
    };
    let f = m.funcs;
    unsafe {
        if f.C_Initialize.unwrap()(std::ptr::null_mut()) != CKR_OK {
            eprintln!("C_Initialize failed");
            return ExitCode::from(2);
        }
    }

    let mut nslots: CK_ULONG = 0;
    let rv = unsafe { f.C_GetSlotList.unwrap()(CK_TRUE, std::ptr::null_mut(), &mut nslots) };
    if rv != CKR_OK || nslots == 0 {
        eprintln!("no usable PKCS#11 slot — pool not fillable yet?");
        unsafe { f.C_Finalize.unwrap()(std::ptr::null_mut()) };
        return ExitCode::from(1);
    }
    let mut slots: Vec<CK_SLOT_ID> = vec![0; nslots as usize];
    let rv = unsafe { f.C_GetSlotList.unwrap()(CK_TRUE, slots.as_mut_ptr(), &mut nslots) };
    if rv != CKR_OK {
        eprintln!("C_GetSlotList failed");
        unsafe { f.C_Finalize.unwrap()(std::ptr::null_mut()) };
        return ExitCode::from(1);
    }

    let mut sess: CK_SESSION_HANDLE = 0;
    let rv = unsafe {
        f.C_OpenSession.unwrap()(
            slots[0],
            CKF_SERIAL_SESSION,
            std::ptr::null_mut(),
            None,
            &mut sess,
        )
    };
    if rv != CKR_OK {
        eprintln!("C_OpenSession failed");
        unsafe { f.C_Finalize.unwrap()(std::ptr::null_mut()) };
        return ExitCode::from(1);
    }

    let mut rfile = match OpenOptions::new()
        .write(true)
        .custom_flags(libc::O_CLOEXEC)
        .open("/dev/random")
    {
        Ok(f) => f,
        Err(e) => {
            eprintln!("open /dev/random: {e}");
            unsafe {
                f.C_CloseSession.unwrap()(sess);
                f.C_Finalize.unwrap()(std::ptr::null_mut());
            }
            return ExitCode::from(1);
        }
    };

    // One contiguous allocation: rand_pool_info header followed by the byte
    // buffer. C_GenerateRandom writes straight into the buf area;
    // RNDADDENTROPY consumes the same memory.
    let header_size = size_of::<RandPoolInfoHeader>();
    let total = header_size + a.nbytes;
    let mut blob = vec![0u8; total];
    {
        let hdr = unsafe { &mut *(blob.as_mut_ptr() as *mut RandPoolInfoHeader) };
        hdr.entropy_count = (a.nbytes * 8) as c_int;
        hdr.buf_size = a.nbytes as c_int;
    }

    signal_reset();
    install_signals(false);

    info!(
        a.quiet,
        "seeding /dev/random: {} B every {:.2} s{}",
        a.nbytes,
        a.interval,
        if a.no_credit { " (no entropy credit)" } else { "" }
    );

    while signal_count() == 0 {
        let buf_ptr = unsafe { blob.as_mut_ptr().add(header_size) };
        let rv = unsafe { f.C_GenerateRandom.unwrap()(sess, buf_ptr, a.nbytes as CK_ULONG) };
        if rv != CKR_OK {
            if signal_count() > 0 {
                break;
            }
            eprintln!("C_GenerateRandom failed; retrying after interval");
            interruptible_sleep(a.interval);
            continue;
        }

        if a.no_credit {
            let buf = unsafe { std::slice::from_raw_parts(buf_ptr, a.nbytes) };
            match rfile.write(buf) {
                Ok(n) => info!(a.quiet, "wrote {n} B (uncredited)"),
                Err(e) => eprintln!("write /dev/random: {e}"),
            }
        } else {
            let r = unsafe {
                libc::ioctl(
                    rfile.as_raw_fd(),
                    RNDADDENTROPY,
                    blob.as_ptr() as *const libc::c_void,
                )
            };
            if r < 0 {
                let err = std::io::Error::last_os_error();
                if err.raw_os_error() == Some(libc::EPERM) {
                    eprintln!(
                        "RNDADDENTROPY: EPERM — falling back to uncredited write \
                         (run as root or with CAP_SYS_ADMIN to credit entropy)"
                    );
                    a.no_credit = true;
                    let buf = unsafe { std::slice::from_raw_parts(buf_ptr, a.nbytes) };
                    if let Err(e) = rfile.write(buf) {
                        eprintln!("write /dev/random: {e}");
                    }
                } else {
                    eprintln!("RNDADDENTROPY: {err}");
                }
            } else {
                info!(a.quiet, "credited {} B ({} bits) to kernel pool", a.nbytes, a.nbytes * 8);
                if !a.no_reseed {
                    let r = unsafe { libc::ioctl(rfile.as_raw_fd(), RNDRESEEDCRNG) };
                    if r < 0 {
                        let err = std::io::Error::last_os_error();
                        if err.raw_os_error() == Some(libc::EPERM) {
                            eprintln!("RNDRESEEDCRNG: EPERM — disabling for the run");
                            a.no_reseed = true;
                        } else {
                            eprintln!("RNDRESEEDCRNG: {err}");
                        }
                    } else {
                        info!(a.quiet, "forced CRNG reseed");
                    }
                }
            }
        }

        interruptible_sleep(a.interval);
    }

    info!(a.quiet, "exiting");

    // Don't leave the byte buffer in memory after exit.
    blob.fill(0);

    unsafe {
        f.C_CloseSession.unwrap()(sess);
        f.C_Finalize.unwrap()(std::ptr::null_mut());
    }
    ExitCode::SUCCESS
}
